lsp/src/Language/LSP/Server/Processing.hs

```diff
@@ -427,17 +435,21 @@ shutdownRequestHandler :: Handler IO Method_Shutdown
 shutdownRequestHandler _req k = do
   k $ Right Null
 
-handleConfigChange :: (m ~ LspM config) => LogAction m (WithSeverity LspProcessingLog) -> TMessage Method_WorkspaceDidChangeConfiguration -> m ()
-handleConfigChange logger req = do
-  parseConfig <- LspT $ asks resParseConfig
-  let s = req ^. L.params . L.settings
-  res <- stateState resConfig $ \oldConfig -> case parseConfig oldConfig s of
-    Left err -> (Left err, oldConfig)
-    Right !newConfig -> (Right (), newConfig)
-  case res of
-    Left err -> do
-      logger <& ConfigurationParseError s err `WithSeverity` Error
-    Right () -> pure ()
+-- | Try to find the configuration section in an object that might represent "all" the settings.
+-- The heuristic we use is to look for a property with the right name, and use that if we find
+-- it. Otherwise we fall back to the whole object.
+-- See Note [LSP configuration]
+lookForConfigSection :: T.Text -> Value -> Value
+lookForConfigSection section (Object o) | Just s' <- o ^. at (fromString $ T.unpack section) = s'
+lookForConfigSection _ o = o
+
+-- | Handle a workspace/didChangeConfiguration request.
+handleDidChangeConfiguration :: (m ~ LspM config) => LogAction m (WithSeverity LspProcessingLog) -> TMessage Method_WorkspaceDidChangeConfiguration -> m ()
+handleDidChangeConfiguration logger req = do
+  section <- LspT $ asks resConfigSection
+  tryChangeConfig (cmap (fmap LspCore) logger) (lookForConfigSection section $ req ^. L.params . L.settings)
+  -- See Note [LSP configuration]
+  requestConfigUpdate (cmap (fmap LspCore) logger)
```

Comment:
I'm confused about why also we need `requestConfigUpdate` if `tryChangeConfig` succeeds.

Reply:
Hmm, that's a good question. I guess if we actually manage to parse out what's in `didChangeConfiguration` then we probably don't need to request new config :thinking: 

Or maybe we should just fully commit to using `workspace/configuration` and not look in the body of `didChangeConfiguration` at all. That seems to be the way that the spec maintainers want us to go...

Reply:
I'm still confused here.. `tryChangeConfig` has updated the config, is there any situation we still need `requestConfigUpdate`?

Reply:
I'll write it out in detail in the doc. Basically, all of the following are "allowable" scenarios:
1. Client supports `workspace/configuration`, sends an empty update in `didChangeConfiguration`. This is what new vscode will do, I believe. Then the first attempt will fail (nothing in `didChange`), and the second will succeed (requesting config).
2. Client does not support `workspace/configuration`, and sends updated config in `didChangeConfiguration`. This is rare, I think, but we should support it. Then the first attempt will succeed and the second attempt will fail (or not even try).

We do get some redundancy in the slightly odd case of a client that supports both ways, but I'm not too worried about that tbh.